Issue 8

Wabri opened the pull request:
Setup Test EmployeeController
Created interface Employee and IEmployeeService
Update Pom
Closes #8

Wabri committed:
SetUp EmployeeController 

created interface IEmployeeService and Employee

Wabri committed:
Update pom with mockito dependecy

LorenzoBettini commented on exercise-core/pom.xml:
```diff
@@ -21,6 +21,12 @@
       <scope>test</scope>
     </dependency>
+    <dependency>
+			<groupId>org.mockito</groupId>
```
Fix the alignment no tabs: 2 spaces

LorenzoBettini commented on exercise-core/src/main/java/com/examples/attsw/exercise/core/model/Employee.java:
```diff
@@ -0,0 +1,5 @@
+package com.examples.attsw.exercise.core.model;
+
+public interface Employee {
```
Create a class instead of an interface

Wabri committed:
Created GetAllEmployeeWhenThereAreNoEmployee

Wabri committed:
update Pom

Wabri committed:
Change Employee to class from interface

Wabri committed:
created test get all employee when there is one Employee

Wabri committed:
tested get Employee by id when there are no employee in list

Wabri committed:
tested get Employee with existing id

LorenzoBettini commented on exercise-core/src/main/java/com/examples/attsw/exercise/core/controller/EmployeeController.java:
```diff
@@ -0,0 +1,37 @@
+
+	public Employee getEmployeeById(String id) {
+		Iterator<Employee> iteratorAllEmployee = iEmployeeService.getEmployees().iterator();
+		while (iteratorAllEmployee.hasNext()) {
```
the retrieval of an employee given its id is part of repository's interface, and simply delegate that to the repository

LorenzoBettini commented on exercise-core/src/main/java/com/examples/attsw/exercise/core/controller/EmployeeController.java:
```diff
@@ -0,0 +1,37 @@
+		return stringEmployee;
+	}
+
+	public Employee getEmployeeById(String id) {
```
must return a String

LorenzoBettini commented on exercise-core/src/main/java/com/examples/attsw/exercise/core/model/Employee.java:
```diff
@@ -0,0 +1,34 @@
+		this.setId(id);
+	}
+
+	public String getName() {
```
only getters and setters if needed

Wabri committed:
made the required changes

Wabri committed:
done some rafactoring

Wabri committed:
done some refactoring in EmployeeController

Wabri committed:
Solve Conflict with pom and Employee

LorenzoBettini commented on exercise-core/src/main/java/com/examples/attsw/exercise/core/controller/EmployeeController.java:
```diff
@@ -0,0 +1,26 @@
+
+public class EmployeeController {
+
+	private IEmployeeService iEmployeeService;
```
there's no need to use "i" in the field name, just use employeeService

LorenzoBettini commented on exercise-core/src/main/java/com/examples/attsw/exercise/core/controller/EmployeeController.java:
```diff
@@ -0,0 +1,26 @@
+	public String getAllEmployee() {
+		String stringEmployee = "";
+		for (Employee employee : iEmployeeService.getEmployees()) {
+			stringEmployee += employee.toString();
```
the issue required that each employee should be in a different line... this does not add any end of line...

LorenzoBettini commented on exercise-core/src/main/java/com/examples/attsw/exercise/core/model/Employee.java:
```diff
@@ -2,14 +2,25 @@
+
+	@Override
+	public String toString() {
+		return getName();
```
also the id should be part of the String. Please have a look at the Java conventions for the toString method, or, simply have Eclipse generate the toString method

LorenzoBettini commented on exercise-core/src/test/java/com/examples/attsw/exercise/core/controller/EmployeeControllerTest.java:
```diff
@@ -0,0 +1,63 @@
+		iEmployeeService = Mockito.mock(IEmployeeService.class);
+		employeeController = new EmployeeController(iEmployeeService);
+		listEmployees = new LinkedList<Employee>();
+		Mockito.when(iEmployeeService.getEmployees()).thenReturn(listEmployees);
```
use import static of Mockito to avoid using Mockito. before static methods

LorenzoBettini commented on exercise-core/src/test/java/com/examples/attsw/exercise/core/controller/EmployeeControllerTest.java:
```diff
@@ -0,0 +1,63 @@
+		Employee employee = newEmployee("nameTest", "idTest");
+		listEmployees.add(employee);
+		assertGetAllEmployee(employee.getName());
+	}
```
when dealing with lists, first test the case when the list is empty (OK), then with one element (OK) and then with two elements which is not tested at all, in fact, the implementation for several elements does not respect the specification of the issue (see the above comment)

LorenzoBettini reviewed (changes requested):
see all the comments. Morever, in the PR you should specify "Closes #15 "

LorenzoBettini commented:
In the PR specify "Closes #8"

Wabri committed:
Review:

Used a static Mockito import, edited toString method, in
EmployeeController renamed iEmployeeService to employeeService

Wabri committed:
Tested Functionality of getAllEmployee with 2 employee in list

Wabri commented on exercise-core/src/test/java/com/examples/attsw/exercise/core/controller/EmployeeControllerTest.java:
```diff
@@ -0,0 +1,80 @@
+		assertEquals(newEmployee.getName(), employeeController.getEmployeeById("1"));
+	}
+	
+	private String extractAllEmployeeStringFromList(List<Employee> allEmployee) {
```
i don't really know if this is the right way

LorenzoBettini replied:
it's OK.

On a side note: since we're using Java 8, such for statements should be replaced by lambdas (also in the SUT); it's easier and more readable to create such string representations with lambdas

Wabri replied:
I'm trying to replace for with this:
`allEmployee.forEach(e -> stringAllEmployee.concat(e.toString()+System.getProperty("line.separator")));`
but it doesn't work at all, look like the concat call doesn't do nothing.

Wabri replied:
It doesn't work because return of concat is a string and I didn't assign it to anything.

LorenzoBettini replied:
In fact that's not the right way to do that :)

LorenzoBettini commented on exercise-core/src/main/java/com/examples/attsw/exercise/core/model/Employee.java:
```diff
@@ -2,14 +2,25 @@
 
 public class Employee {
 
+	private String name;
```
Let's take the opportunity to put "id" as the first field; toString should be changed accordingly

LorenzoBettini commented on exercise-core/src/main/java/com/examples/attsw/exercise/core/model/Employee.java:
```diff
@@ -2,14 +2,25 @@
 	private String id;
 
-	public Employee(String id) {
+	public Employee(String name, String id) {
```
update constructor according to the order of the field (see above); use Eclipse refactoring.

LorenzoBettini commented on exercise-core/src/main/java/com/examples/attsw/exercise/core/model/Employee.java:
```diff
@@ -2,14 +2,25 @@
+
+	@Override
+	public String toString() {
+		return "Employee [name=" + name + ", id=" + id + "]";
```
same here concerning the order

Wabri commented:
Closes #8

LorenzoBettini commented on exercise-core/src/test/java/com/examples/attsw/exercise/core/controller/EmployeeControllerTest.java:
```diff
@@ -0,0 +1,80 @@
+	}
+	
+	@Test
+	public void testGetAllEmployeeWhenThereIsTwoEmployee() {
```
ThereAre not ThereIs

Wabri committed:
Changed fields order in Employee class

Wabri committed:
Solve Conflict

Wabri committed:
replaced for each with lambda expression in class EmployeeController for
method getAllEmployee

LorenzoBettini commented on exercise-core/src/main/java/com/examples/attsw/exercise/core/repository/mongo/MongoRepository.java:
```diff
@@ -16,6 +15,7 @@
 
 public class MongoRepository implements Repository {
 
+	@SuppressWarnings("unused")
```
why?!

LorenzoBettini commented on exercise-core/src/test/java/com/examples/attsw/exercise/core/controller/EmployeeControllerTest.java:
```diff
@@ -0,0 +1,81 @@
+		assertEquals(newEmployee.getName(), employeeController.getEmployeeById("1"));
+	}
+
+	private String extractAllEmployeeStringFromList(List<Employee> allEmployee) {
```
should be "allEmployees"

LorenzoBettini commented on exercise-core/src/test/java/com/examples/attsw/exercise/core/controller/EmployeeControllerTest.java:
```diff
@@ -0,0 +1,81 @@
+	}
+
+	private void assertGetAllEmployee(String expected) {
+		String allEmployee = employeeController.getAllEmployee();
```
should be "allEmployees"

LorenzoBettini reviewed (changes requested):
just a few things left to do

Wabri committed:
solve different spelling problem

LorenzoBettini commented on exercise-core/src/main/java/com/examples/attsw/exercise/core/controller/EmployeeController.java:
```diff
@@ -0,0 +1,26 @@
+		this.employeeService = iEmployeeService;
+	}
+
+	public String getAllEmployee() {
```
OOPS... my mistake I called that method getAllEmployee in the original issue (which I now updated): of course this should be getAllEmployees, so please refactor also your code

Wabri committed:
Hope now it's fine
